dataland-backend-utils/src/main/kotlin/org/dataland/datalandbackendutils/apikey/ApiKeyUtility.kt

```diff
@@ -0,0 +1,119 @@
+package org.dataland.datalandbackendutils.apikey
+
+import org.dataland.datalandbackendutils.exceptions.ApiKeyFormatException
+import org.dataland.datalandbackendutils.utils.EncodingUtils.calculateCrc32Value
+import org.dataland.datalandbackendutils.utils.EncodingUtils.decodeFromBase64
+import org.dataland.datalandbackendutils.utils.EncodingUtils.encodeToBase64
+import org.springframework.security.crypto.argon2.Argon2PasswordEncoder
+import java.security.SecureRandom
+import java.util.HexFormat
+
+/**
+ * This class should be used to validate that a given api key has the required format and the correct checksum
+ * before a request is sent to /api-keys/validateApiKey to prevent unnecessary traffic
+ */
+class ApiKeyUtility {
+
+    companion object {
+        private const val keyByteLength = 40
+    }
+
+    private val numberOfUnderscoreDelimitersExpectedInApiKey = 2
+
+    private val regexFor80HexCharacters = Regex("^[a-fA-F0-9]{${keyByteLength * 2}}\$")
```

Comment:
Wieso ist das $-Zeichen hier eigentlich escaped? Endet der String immer mit "$"

Reply:
Das Dollarzeichen muss hier ja als anchor für das Ende der line zwingend stehen.
Wenn ich es nicht escape, hätte ich die Sorge, dass Kotlin das Dollarzeichen nicht literally in den String steckt, sondern auflöst.
Irgendwo bei regex101 hatte ich gelesen "Achtung, escape das Dollarzeichen, wenn du die Regex im Code verwendest".

Jetzt habe ich aber gerade im Kotlin Playground gesehen, dass das Dollarzeichen literal verwendet wird, wenn darauf keine expression folgt.

Also:

println("hallo$") => hallo$

println("hallo${ }" ) =>   hallo

D.h. ich müsste hier nicht zwingend escapen.
Soll ich es also nicht escapen?  Schaden tut es ja nicht, und es zeigt zumindest dem Code-Leser, dass man hier explizit ein tatsächliches Dollarzeichen haben möchte.

Reply:
Also wenn der String den man matchen möchte immer mit "$" enden muss also zum Beispiel "hallo$" dann lass das escape so wie es ist. Ich war nur etwas irritiert weil es oft auch solche Konstruktionen als Regex gibt "^[0-9][0-9]$" womit jetzt alle zweistelligen Zahlen matchen. Also wo man einen exakten match erzwingen will durch beginnt mit, endet mit.

Reply:
Nein, der String an dieser Stelle endet niemals mit $.
Das Dollarzeichen soll hier wie du sagst einen exakten match erzwingen, und keinen submatch.

Reply:
Wenn man "hallo$", aber nicht "hallo$1" matchen machte, dann wäre die Regex dafür z.B.:
"[a-z]+\\\$", wenn man aber "hallo" matchen möchte, und nicht "hallo1", dann wäre eine Regex:
"[a-z]+\$"

Reply:
alles klar